src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpBindJoinSPARQLwithUNION.java

```diff
@@ -0,0 +1,80 @@
+package se.liu.ida.hefquin.engine.queryplan.executable.impl.ops;
+
+import java.util.Iterator;
+import java.util.Set;
+
+import org.apache.jena.graph.Node;
+import org.apache.jena.sparql.algebra.Op;
+import org.apache.jena.sparql.algebra.op.OpFilter;
+import org.apache.jena.sparql.algebra.op.OpTriple;
+import org.apache.jena.sparql.algebra.op.OpUnion;
+import org.apache.jena.sparql.core.Var;
+import org.apache.jena.sparql.engine.binding.Binding;
+import org.apache.jena.sparql.expr.E_Equals;
+import org.apache.jena.sparql.expr.E_LogicalAnd;
+import org.apache.jena.sparql.expr.Expr;
+import org.apache.jena.sparql.expr.ExprVar;
+import org.apache.jena.sparql.expr.nodevalue.NodeValueNode;
+
+import se.liu.ida.hefquin.engine.data.SolutionMapping;
+import se.liu.ida.hefquin.engine.data.impl.SolutionMappingUtils;
+import se.liu.ida.hefquin.engine.federation.SPARQLEndpoint;
+import se.liu.ida.hefquin.engine.federation.access.SPARQLRequest;
+import se.liu.ida.hefquin.engine.federation.access.SolMapsResponse;
+import se.liu.ida.hefquin.engine.federation.access.impl.req.SPARQLRequestImpl;
+import se.liu.ida.hefquin.engine.query.SPARQLGraphPattern;
+import se.liu.ida.hefquin.engine.query.TriplePattern;
+import se.liu.ida.hefquin.engine.query.impl.QueryPatternUtils;
+import se.liu.ida.hefquin.engine.query.impl.SPARQLGraphPatternImpl;
+import se.liu.ida.hefquin.engine.queryproc.ExecutionContext;
+
+public class ExecOpBindJoinSPARQLwithUNION extends ExecOpGenericBindJoin<TriplePattern,SPARQLEndpoint> {
+
+	private Set<Var> varsInTP;
+
+	public ExecOpBindJoinSPARQLwithUNION( final TriplePattern query, final SPARQLEndpoint fm) {
+		super(query, fm);
+		varsInTP = QueryPatternUtils.getVariablesInPattern(query);
+	}
+
+	@Override
+	protected Iterable<? extends SolutionMapping> fetchSolutionMappings( final Iterable<SolutionMapping> solMaps,
+			final ExecutionContext execCxt) {
+		final Op op = createUnion(solMaps);
+		final SPARQLGraphPattern pattern = new SPARQLGraphPatternImpl(op);
+		final SPARQLRequest request = new SPARQLRequestImpl(pattern);
+		final SolMapsResponse response = execCxt.getFederationAccessMgr().performRequest(request, fm);
+		return response.getSolutionMappings();
+	}
+
+	protected Op createUnion(final Iterable<SolutionMapping> solMaps) {
+		Op tp = new OpTriple(query.asJenaTriple());
+		if (varsInTP.isEmpty()) return tp;
+		Op union = null;
+		for ( final SolutionMapping s : solMaps) {
+			final Binding b = SolutionMappingUtils.restrict(s.asJenaBinding(), varsInTP);
+			if (b.size() == 0) continue;
```

Comment:
This is wrong. Every solution mapping `s` that does not have any variables in common with the triple pattern is a join partner for *every* solution mapping that can be produced for the triple pattern. Hence, as soon as the is one such `s`, the triple pattern has to be included in the request without any additional filter. In fact, in this case, there is no need to send a UNION-based request at all; in contrast, the triple pattern can be sent just by itself. Hence, I believe that the correct behavior here is the following.
```suggestion
			// If the current solution mapping does not have any variables in common with
			// the triple pattern of this operator, then every matching triple is a join partner
			// for the current solution mapping. Hence, in this case, we may simply retrieve
			// all matching triples (i.e., no need for putting together the UNION pattern).
			if (b.size() == 0) return tp;
```

Reply:
On the meantime I have committed this change. However, I ask if the fact that one of the restricted solution mappings has size 0 implies that every solution mapping will have a size 0. I believe the implication is not correct, thus I would have to add `tp` to the union when the binding size is 0 instead of returning just the triple pattern.

Reply:
In general we cannot assume that implication, I believe. On the other hand, there are cases where the implication holds. However, in these cases, this can be figured out already before the query execution process; namely, based on the ExpectedVariables of the input operator together with the variables of the given triple pattern. 

Having said that, no matter what, it will always be correct to simply return the triple pattern in all of these cases. The reason is that you use this method here as a basis for the fetchSolutionMappings method above, and that method is just meant to fetch all solution mappings from the SPARQL endpoint that are compatible with at least one of the given solution mappings. Sending only the triple pattern achieves that. The actual join happens then in the superclass after fetchSolutionMappings is called.

Reply:
alright!